File: src/main/java/personal/alexgarland/helloworldmvc/service/AbstractDatabaseEmployeeRepository.java
package personal.alexgarland.helloworldmvc.service;

import java.io.FileInputStream;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.List;
import java.util.Properties;

import personal.alexgarland.helloworldmvc.model.Employee;


public abstract class AbstractDatabaseEmployeeRepository implements IEmployeeRepository {

	private final Properties dbProperties;

	protected abstract String getDbPropertiesFileName();
	
	public AbstractDatabaseEmployeeRepository() {
		ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
		String path = classLoader.getResource(getDbPropertiesFileName()).getPath();
		Properties properties = new Properties();
		try {
			properties.load(new FileInputStream(path));
		}
		catch (IOException ex) {
			throw new RuntimeException(ex);
		}
		this.dbProperties = properties;
	}
	
	private Connection getConnection() throws SQLException {
		return DriverManager.getConnection(dbProperties.getProperty("url"), dbProperties);
	}
	
	@FunctionalInterface
	private interface PreparedStatementFunction<R> {
		R apply(PreparedStatement ps) throws SQLException;
	}

	private <A> A withPreparedStatement(String sql, PreparedStatementFunction<A> function) {
		 try (Connection con = getConnection();
				 PreparedStatement ps = con.prepareStatement(sql);) {		 
			 return function.apply(ps);
			 }
		 catch (SQLException ex) {
			 throw new RuntimeException(ex.getMessage());
			 }
	}
	
	protected abstract Employee addEmployeeWithPreparedStatement(PreparedStatement ps, Employee e) throws SQLException;
	protected abstract String getAddEmployeeQuery();

	@Override
	public Employee addEmployee(Employee e) {
		return withPreparedStatement(getAddEmployeeQuery(), (ps) -> addEmployeeWithPreparedStatement(ps, e));
	}
	
	protected abstract List<Employee> getEmployeeListWithPreparedStatement(PreparedStatement ps) throws SQLException;
	protected abstract String getGetEmployeeListQuery();

	@Override
	public List<Employee> getEmployeeList() {
		return withPreparedStatement(getGetEmployeeListQuery(), (ps) -> getEmployeeListWithPreparedStatement(ps));
	}
	
	protected abstract Integer deleteEmployeeWithPreparedStatement(PreparedStatement ps, int employeeId) throws SQLException;
	protected abstract String getDeleteQuery();

	@Override
	public void deleteEmployeeById(int employeeId) {
		withPreparedStatement(getDeleteQuery(),(ps) -> deleteEmployeeWithPreparedStatement(ps, employeeId));
	}
	
	protected abstract Integer updateEmployeeWithPreparedStatement(PreparedStatement ps, Employee e) throws SQLException;
	protected abstract String getUpdateQuery();

	@Override
	public void updateEmployee(Employee e) {
		withPreparedStatement(getUpdateQuery(), (ps) -> updateEmployeeWithPreparedStatement(ps, e));
	}
	
	protected abstract Employee getEmployeeByIdWithPreparedStatement(PreparedStatement ps, int employeeId) throws SQLException;
	protected abstract String getGetEmployeeByIdQuery();
	
	@Override
	public Employee getEmployeeById(int employeeId) {
		return withPreparedStatement(getGetEmployeeByIdQuery(), (ps) -> getEmployeeByIdWithPreparedStatement(ps, employeeId));
	}
	
}

File: src/main/kotlin/personal/alexgarland/helloworldmvc/service/AbstractDatabaseEmployeeRepository.kt
package personal.alexgarland.helloworldmvc.service

import personal.alexgarland.helloworldmvc.model.Employee
import java.io.FileInputStream
import java.io.IOException
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement
import java.sql.SQLException
import java.util.*

abstract class AbstractDatabaseEmployeeRepository : IEmployeeRepository {

    private val dbProperties: Properties

    protected abstract val dbPropertiesFileName: String

    private val connection: Connection
        @Throws(SQLException::class)
        get() = DriverManager.getConnection(dbProperties.getProperty("url"), dbProperties)

    protected abstract val addEmployeeQuery: String

    protected abstract val getEmployeeListQuery: String

    protected abstract val deleteQuery: String

    protected abstract val updateQuery: String

    protected abstract val getEmployeeByIdQuery: String

    init {
        val classLoader = Thread.currentThread().contextClassLoader
        val path = classLoader.getResource(dbPropertiesFileName)!!.path
        val properties = Properties()
        try {
            properties.load(FileInputStream(path))
        } catch (ex: IOException) {
            throw RuntimeException(ex)
        }

        this.dbProperties = properties
    }

    private fun <A> withPS(sql: String, function: (PreparedStatement) -> A): A {
        try {
            connection.use { con -> con.prepareStatement(sql).use { ps -> return function(ps) } }
        } catch (ex: SQLException) {
            throw RuntimeException(ex.message)
        }
    }

    @Throws(SQLException::class)
    protected abstract fun addEmployeeWithPS(ps: PreparedStatement, e: Employee): Employee

    override fun addEmployee(e: Employee): Employee {
        return withPS(addEmployeeQuery, { ps -> addEmployeeWithPS(ps, e) })
    }

    @Throws(SQLException::class)
    protected abstract fun getEmployeeListWithPS(ps: PreparedStatement): List<Employee>

    override val employeeList: List<Employee>
        get() {
            return withPS(getEmployeeListQuery, { ps -> getEmployeeListWithPS(ps) })
        }

    @Throws(SQLException::class)
    protected abstract fun deleteEmployeeWithPS(ps: PreparedStatement, employeeId: Int): Int

    override fun deleteEmployeeById(employeeId: Int) {
        withPS<Int>(deleteQuery, { ps -> deleteEmployeeWithPS(ps, employeeId) })
    }

    @Throws(SQLException::class)
    protected abstract fun updateEmployeeWithPS(ps: PreparedStatement, e: Employee): Int

    override fun updateEmployee(e: Employee) {
        withPS(updateQuery, { ps -> updateEmployeeWithPS(ps, e) })
    }

    @Throws(SQLException::class)
    protected abstract fun getEmployeeByIdWithPS(ps: PreparedStatement, employeeId: Int): Employee

    override fun getEmployeeById(employeeId: Int): Employee {
        return withPS(getEmployeeByIdQuery, { ps -> getEmployeeByIdWithPS(ps, employeeId) })
    }

}